base_riskscore/code/utils.R

```diff
@@ -0,0 +1,537 @@
+# get the CV-AUC for a single learner's predicted values
+# @param preds the fitted values
+# @param Y the outcome
+# @param scale what scale should the IPCW correction be applied on?
+#              Can help with numbers outside of (0, 1)
+#             ("identity" denotes the identity scale;
+#              "logit" means that AUC is transformed to the logit scale,
+#              correction is applied, and then back-transformed)
+# @param weights the inverse probability of censoring weights
+# @param C the indicator of being observed in phase 2 (1) or not (0)
+# @param Z a matrix of predictors observed on all participants in phase 1
+#          (can include the outcome)
+# @param ... other arguments to measure_auc, but should include at least:
+#            a library of learners (using arg "SL.library")
+#            and may include control parameters for the super learner
+#            (e.g., cvControl = list(V = 5)
+#             for 5-fold cross-validated super learner)
+one_auc <- function(preds, Y, full_y = NULL, scale = "identity",
+                    weights = rep(1, length(Y)), C = rep(1, length(Y)),
+                    Z = NULL, ...) {
+  auc_lst <- vimp::measure_auc(
+    fitted_values = preds, y = Y, full_y = full_y, C = C, Z = Z,
+    ipc_weights = weights,
+    ipc_fit_type = "SL", ...
+  )
+  data.frame(auc = auc_lst$point_est, eif = auc_lst$eif)
+}
+
+# get the cross-fitted CV-AUC for a single learner's predicted values
+# @param preds the fitted values
+# @param Y the outcome
+# @param folds the different cv folds that the learner was evaluated on
+# @param scale what scale should the IPCW correction be applied on?
+#              Can help with numbers outside of (0, 1)
+#             ("identity" denotes the identity scale;
+#              "logit" means that AUC is transformed to the logit scale,
+#              correction is applied, and then back-transformed)
+# @param weights the inverse probability of censoring weights
+# @param C the indicator of being observed in phase 2 (1) or not (0)
+# @param Z a matrix of predictors observed on all participants in phase 1
+#          (can include the outcome)
+# @param ... other arguments to measure_auc, but should include at least:
+#            a library of learners (using arg "SL.library")
+#            and may include control parameters for the super learner
+#            (e.g., cvControl = list(V = 5)
+#             for 5-fold cross-validated super learner)
+cv_auc <- function(preds, Y, folds, scale = "identity",
+                   weights = rep(1, length(Y)), C = rep(1, length(Y)),
+                   Z = NULL, ...) {
+  V <- length(folds)
+  folds_numeric <- get_cv_sl_folds(folds)
+  if (is.null(Z)) {
+    folds_z <- folds_numeric
+  } else {
+    folds_z <- c(folds_numeric, sample(seq_len(V), nrow(Z) - length(folds_numeric),
+      replace = TRUE
+    ))
+  }
+  ests_eifs <- lapply(as.list(1:V), function(v) {
```

Comment:
```suggestion
  ests_eifs <- lapply(as.list(seq_len(V)), function(v) {
```
`seq_len()` is generally safer than `1:V`, since certain values of V can cause `1:V` to evaluate in unexpected ways, whereas `seq_len(V)` will always return the sequence from 1 to the value of a scalar V